src/wp-includes/sitemaps/class-wp-sitemaps-provider.php

```diff
@@ -0,0 +1,190 @@
+<?php
+/**
+ * Sitemaps: WP_Sitemaps_Provider class
+ *
+ * This class is a base class for other sitemap providers to extend and contains shared functionality.
+ *
+ * @package WordPress
+ * @subpackage Sitemaps
+ * @since 5.5.0
+ */
+
+/**
+ * Class WP_Sitemaps_Provider.
+ *
+ * @since 5.5.0
+ */
+abstract class WP_Sitemaps_Provider {
+	/**
+	 * Provider name.
+	 *
+	 * This will also be used as the public-facing name in URLs.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @var string
+	 */
+	protected $name = '';
+
+	/**
+	 * Object type name (e.g. 'post', 'term', 'user').
+	 *
+	 * @since 5.5.0
+	 *
+	 * @var string
+	 */
+	protected $object_type = '';
+
+	/**
+	 * Gets a URL list for a sitemap.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param int    $page_num       Page of results.
+	 * @param string $object_subtype Optional. Object subtype name. Default empty.
+	 * @return array $url_list Array of URLs for a sitemap.
+	 */
+	abstract public function get_url_list( $page_num, $object_subtype = '' );
+
+	/**
+	 * Gets the max number of pages available for the object type.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param string $object_subtype Optional. Object subtype. Default empty.
+	 * @return int Total number of pages.
+	 */
+	abstract public function get_max_num_pages( $object_subtype = '' );
+
+	/**
+	 * Gets data about each sitemap type.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @return array List of sitemap types including object subtype name and number of pages.
+	 */
+	public function get_sitemap_type_data() {
+		$sitemap_data = array();
+
+		$object_subtypes = $this->get_object_subtypes();
+
+		// If there are no object subtypes, include a single sitemap for the
+		// entire object type.
+		if ( empty( $object_subtypes ) ) {
+			$sitemap_data[] = array(
+				'name'  => '',
+				'pages' => $this->get_max_num_pages(),
+			);
+			return $sitemap_data;
+		}
+
+		// Otherwise, include individual sitemaps for every object subtype.
+		foreach ( $object_subtypes as $object_subtype_name => $data ) {
+			$object_subtype_name = (string) $object_subtype_name;
+
+			$sitemap_data[] = array(
+				'name'  => $object_subtype_name,
+				'pages' => $this->get_max_num_pages( $object_subtype_name ),
+			);
+		}
+
+		return $sitemap_data;
+	}
+
+	/**
+	 * Lists sitemap pages exposed by this provider.
+	 *
+	 * The returned data is used to populate the sitemap entries of the index.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @return array List of sitemaps.
+	 */
+	public function get_sitemap_entries() {
+		$sitemaps = array();
+
+		$sitemap_types = $this->get_sitemap_type_data();
+
+		foreach ( $sitemap_types as $type ) {
+			for ( $page = 1; $page <= $type['pages']; $page ++ ) {
+				$loc        = $this->get_sitemap_url( $type['name'], $page );
+				$sitemap_entry = array(
+					'loc' => $this->get_sitemap_url( $type['name'], $page ),
```

Comment:
Why not `'loc' => $loc,` considering that variable was defined just above?

Otherwise, it seems that `$loc` is unused?

Reply:
this was caught after the PR was merged into core, and corrected in [[48080]](https://core.trac.wordpress.org/changeset/48080).